pkg/reconciler/source/crd/crd.go

```diff
@@ -0,0 +1,308 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package crd
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/cache"
+	"knative.dev/eventing/pkg/logging"
+	"knative.dev/eventing/pkg/reconciler/source/duck"
+	"knative.dev/pkg/configmap"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/controller"
+
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/client/listers/apiextensions/v1beta1"
+	"knative.dev/eventing/pkg/reconciler"
+)
+
+const (
+	// Name of the corev1.Events emitted from the Source CRDs reconciliation process.
+	sourceCRDReconcileFailed = "SourceCRDReconcileFailed"
+
+	finalizerName = controllerAgentName
+)
+
+type runningController struct {
+	controller *controller.Impl
+	cancel     context.CancelFunc
+}
+
+// Reconciler implements controller.Reconciler for Source CRDs resources.
+type Reconciler struct {
+	*reconciler.Base
+
+	// Listers index properties about resources
+	crdLister apiextensionsv1beta1.CustomResourceDefinitionLister
+	// apiExtensionsClientSet used to update finalizers on CRDs.
+	apiExtensionsClientSet clientset.Interface
+
+	ogctx context.Context
+	ogcmw configmap.Watcher
+
+	// controllers keeps a map for GVR to dynamically created controllers.
+	controllers map[schema.GroupVersionResource]runningController
+
+	// Synchronization primitives
+	lock     sync.RWMutex
+	onlyOnce sync.Once
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+func (r *Reconciler) Reconcile(ctx context.Context, key string) error {
+
+	// Create controllers map only once.
+	r.onlyOnce.Do(func() {
+		r.controllers = make(map[schema.GroupVersionResource]runningController)
+	})
+
+	// Convert the namespace/name string into a distinct namespace and name.
+	_, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		logging.FromContext(ctx).Error("invalid resource key")
+		return nil
+	}
+
+	// Get the CRD resource with this name.
+	original, err := r.crdLister.Get(name)
+	if apierrs.IsNotFound(err) {
+		// The resource may no longer exist, in which case we stop processing.
+		logging.FromContext(ctx).Error("CRD key in work queue no longer exists")
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy.
+	crd := original.DeepCopy()
+
+	var reconcileErr error
+	if crd.GetDeletionTimestamp().IsZero() {
+		// add the finalizer
+		if crd, err = r.addFinalizer(ctx, crd); err != nil {
```

Comment:
do not edit the CRDs.

Reply:
ok will remove and this will make it easier